[HLSL][DXIL][SPIRV] Create llvm dot intrinsic and use for HLSL

pow2clk opened the pull request:
Per https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294
`dot` should be an LLVM intrinsic. This adds the llvm intrinsics
and updates HLSL builtin codegen to emit them.

Removed some stale comments that gave the obsolete impression that
type conversions should be expected to match overloads.

With dot moving into an LLVM intrinsic, the lowering to dx-specific
operations doesn't take place until DXIL intrinsic expansion. This
moves the introduction of arity-specific DX opcodes to DXIL
intrinsic expansion.

The new LLVM integer intrinsics replace the previous dx intrinsics.
This updates the DXIL intrinsic expansion code and tests to use and
expect the new integer intrinsics and the flattened DX floating
vector size variants only after op lowering.

Use the new LLVM dot intrinsics to build SPIRV instructions.
This involves generating multiply and add operations for integers
and the existing OpDot operation for floating point. This includes
adding some generic opcodes for signed, unsigned and floats.
These require updating an existing test for all such opcodes.

New tests for generating SPIRV float and integer dot intrinsics are
added as well.

Fixes #88056

Commit:
Create llvm dot intrinsic

Per https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294
`dot` should be an LLVM intrinsic. This adds the llvm intrinsics
and updates HLSL builtin codegen to emit them.

Removed some stale comments that gave the obsolete impression that
type conversions should be expected to match overloads.

With dot moving into an LLVM intrinsic, the lowering to dx-specific
operations doesn't take place until DXIL intrinsic expansion. This
moves the introduction of arity-specific DX opcodes to DXIL
intrinsic expansion.

Part of #88056

Commit:
Add SPIRV generation for HLSL dot

Use the new LLVM dot intrinsics to build SPIRV instructions.
This involves generating multiply and add operations for integers
and the existing OpDot operation for floating point. This includes
adding some generic opcodes for signed, unsigned and floats.
These require updating an existing test for all such opcodes.

New tests for generating SPIRV float and integer dot intrinsics are
added as well.

Fixes #88056

Commit:
Update DX intrinsic expansion for new llvm intrinsics

The new LLVM integer intrinsics replace the previous dx intrinsics.
This updates the DXIL intrinsic expansion code and tests to use and
expect the new integer intrinsics and the flattened DX floating
vector size variants only after op lowering.

Part of #88056

pow2clk commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18528,37 +18520,38 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
         E->getArg(1)->getType()->getAs<VectorType>();
-    // A HLSLVectorTruncation should have happend
+
+    assert(VecTy0->getElementType() == VecTy1->getElementType() &&
```
Switched to clang types to match signedness of integers

pow2clk commented on llvm/include/llvm/Target/GenericOpcodes.td:
```diff
@@ -1057,6 +1057,27 @@ def G_FTANH : GenericInstruction {
+}
+
+/// Unsigned integer vector dot product
+def G_UDOTPROD : GenericInstruction {
```
The unweildy names are because G_UDOT and G_SDOT clashed with existing AArch64 intrinsics that take three arguments as one is an accumulated inout parameter. https://github.com/llvm/llvm-project/blob/908c89e04b6019bdb08bb5f1c861af42046db623/llvm/lib/Target/AArch64/AArch64InstrGISel.td#L254

pow2clk commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -659,7 +659,7 @@ def Dot3 :  DXILOp<55, dot3> {
 def Dot4 :  DXILOp<56, dot4> {
   let Doc = "dot product of two float vectors Dot(a,b) = a[0]*b[0] + ... + "
-            "a[n]*b[n] where n is between 0 and 3";
+            "a[n]*b[n] where n is 0 to 3 inclusive";
```
Just something incidental as I found these descriptions misleading since the only numbers "between" 0 and 3 are 1 and 2.

pow2clk commented on clang/test/CodeGenHLSL/builtins/dot.hlsl:
```diff
@@ -7,155 +7,155 @@
+// NATIVE_HALF: %dot = call half @llvm.fdot.v2f16(<2 x half> %0, <2 x half> %1)
+// NATIVE_HALF: ret half %dot
+// NO_HALF: %dot = call float @llvm.fdot.v2f32(<2 x float> %0, <2 x float> %1)
+// NO_HALF: ret float %dot
```
Before this, the only change was in the temp names since they are no longer dx-exclusive ops. Here and hereafter, for floating-point values, we no longer lower to the vector-size-specific ops until DXIL intrinsic expansion, so these have a more generic form here.

farzonl replied:
while it is no longer a dx exclusive op we are  still  emitting the intrinsic in an hlsl specific code  section. so instead of using `%dot = ...`  we should use `%hlsl.dot`.

pow2clk commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -380,6 +383,20 @@ bool SPIRVInstructionSelector::spvSelect(Register ResVReg,
+        .addUse(I.getOperand(1).getReg())
+        .addUse(I.getOperand(2).getReg())
+        .constrainAllUses(TII, TRI, RBI);
+  }
```
There is a similar implementation here: https://github.com/llvm/llvm-project/blob/a0241e710fcae9f439e57d3a294b1ace97c6906c/llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp#L1524 , but I'm not sure if they are mergeable and this is what was discussed.

farzonl replied:
I'm not sure myself. This might be a question for @Keenuts, @sudonatalie, or @VyacheslavLevytskyy.

VyacheslavLevytskyy replied:
I think trying to merge those two `SPIRV::OpDot`'s would not simplify or improve anything, let's keep as is.

pow2clk reviewed (commented):
The three commits are independently committable, but this is the grouping @farzonl and I agreed on. Reviewing them individually still might make this easier: 

1. Create llvm dot intrinsic (6fde4bc98d0156024cf7acc27e2e986b9bec3993)
2. Update DX intrinsic expansion for new llvm intrinsics (7ca6bc5940321c18f5634bb960fa795366097e45)
3. Add SPIRV generation for HLSL dot (490c0c05c5762a622d037b472c85234ce3f39c96)

tschuett commented:
Please advertise your achievements: https://github.com/llvm/llvm-project/blob/main/llvm/docs/GlobalISel/GenericOpcode.rst
Please add the intrinsics with semantics to the langref: https://github.com/llvm/llvm-project/blob/main/llvm/docs/LangRef.rst
@nikic will review
Please protect against misuse of your achievements: 
https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MachineVerifier.cpp

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18528,37 +18520,38 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
     if (!T0->isVectorTy() && !T1->isVectorTy()) {
       if (T0->isFloatingPointTy())
-        return Builder.CreateFMul(Op0, Op1, "dx.dot");
+        return Builder.CreateFMul(Op0, Op1, "dot");
```
`hlsl.dot`

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18528,37 +18520,38 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
 
       if (T0->isIntegerTy())
-        return Builder.CreateMul(Op0, Op1, "dx.dot");
+        return Builder.CreateMul(Op0, Op1, "dot");
```
`hlsl.dot`

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18470,22 +18470,14 @@ llvm::Value *CodeGenFunction::EmitScalarOrConstFoldImmArg(unsigned ICEArguments,
+  if (QT->isFloatingType())
+    return Intrinsic::fdot;
+  if (QT->isSignedIntegerType())
+    return Intrinsic::sdot;
```
I don't think we want to do an llvm dot intrinsic for integers. I'm pretty sure the RFC just covered the float case. I would instead do `CGM.getHLSLRuntime().getSDotIntrinsic()`

farzonl replied:
hmm rereading kparzysz  posts in the rfc and seeing aarch64 wants integer dot seems like my comment here isn't correct. I still feel weird about it because our usages of it don't lower to a specific opcode in either the SPIRV or DXIL backends.

pow2clk replied:
Justin's proposal didn't say explicitly, but linked to the [HLSL documentation ](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-dot), which explicitly includes integers. In the discussion, there was an [explicit request](https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294/3) for integer versions that no one disapproved of there.

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18470,22 +18470,14 @@ llvm::Value *CodeGenFunction::EmitScalarOrConstFoldImmArg(unsigned ICEArguments,
+  if (QT->isSignedIntegerType())
+    return Intrinsic::sdot;
+  assert(QT->isUnsignedIntegerType());
+  return Intrinsic::udot;
```
`CGM.getHLSLRuntime().getUDotIntrinsic()`

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18528,37 +18520,38 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
-                               VecTy0->getNumElements()),
-        ArrayRef<Value *>{Op0, Op1}, nullptr, "dx.dot");
+        getDotProductIntrinsic(VecTy0->getElementType()),
+        ArrayRef<Value *>{Op0, Op1}, nullptr, "dot");
```
`hlsl.dot`

farzonl commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -1045,6 +1045,15 @@ let IntrProperties = [IntrNoMem, IntrSpeculatable, IntrWillReturn] in {
+  def int_sdot : Intrinsic<[LLVMVectorElementType<0>],
+                           [llvm_anyint_ty, LLVMScalarOrSameVectorWidth<0, LLVMVectorElementType<0>>],
+                           [IntrNoMem, IntrWillReturn, Commutative] >;
+  def int_fdot : Intrinsic<[LLVMVectorElementType<0>],
```
I had a cleanup I wanted to do to swith to `DefaultAttrsIntrinsic` I don't remember what the full list of defaut attributes are at the moment. but  i'm pretty sure only one we would need to pass in if we switch is `Commutative`.

pow2clk replied:
The default properties seem to be 

* `IntrNoCallBack`
* `IntrNoSync`
* `IntrNoFree`
* `IntrWillReturn`

It seems we need to set `IntrNoMem` too then.

farzonl commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -70,31 +71,57 @@ static bool expandAbs(CallInst *Orig) {
-  Intrinsic::ID MadIntrinsic = DotIntrinsic == Intrinsic::dx_sdot
-                                   ? Intrinsic::dx_imad
-                                   : Intrinsic::dx_umad;
+static bool expandDotIntrinsic(CallInst *Orig, Intrinsic::ID DotIntrinsic) {
```
for simplicity I would keep a seperate  `expandIntegerDot` and `expandFloatDot`. Doing it this way is a little weird considering you already have a conditional to seperate behavior via the switch cases.  Then you merge back only to seperate out the behaviors again via contional `if (EltTy->isIntegerTy()) {`

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1366,6 +1383,67 @@ bool SPIRVInstructionSelector::selectRsqrt(Register ResVReg,
       .constrainAllUses(TII, TRI, RBI);
 }
 
+// Since there is no integer dot implementation, expand by piecewise multiplying
```
actually I see a `OpSDot`, `OpUDot`, and `OpSUDot` (for mixes signedness)  in the [spec](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html). Maybe we just don't use these in DXC's spirv generation. I'm fine with sticking with what DXC does we should at least mark the comment accordingly.

pow2clk replied:
Those are fairly recent SPIRV extensions. I didn't think incorporating them was within scope.

davemgreen commented:
AArch64 has a udot and sdot instruction (and a usdot instruction). They perform a "partial" reduction though, producing a v4i32 from two v16i8 inputs. We would like to use those from the vectorizer and have recently added a partial-reduction intrinsic, but doing it with a higher level intrinsic might be a little nicer.

It would seem like a "udot" can be represented already as `vecreduce.add(mul(zext, zext))`, and fdot is simpler still. Is there any particular reason to add a new intrinsic for it if it is already representable as a vecreduce? And it would feel like a shame if it couldn't be used with the actual AArch64 instructions. 

@SamTebbs33 @NickGuy-Arm FYI.

efriedma-quic commented:
Missing LangRef changes.

Commit:
Respond to feedback

Rename dot-related ops to hlsl.dot

Add documentation for new Genereic opcodes and llvm intrinsics

Use DefaultAttrsIntrinsic to define new llvm intrinsics

Split DXIL instruction expansion for integer and float dot products

SamTebbs33 commented:
> > It would seem like a "udot" can be represented already as `vecreduce.add(mul(zext, zext))`, and fdot is simpler still. Is there any particular reason to add a new intrinsic for it if it is already representable as a vecreduce? And it would feel like a shame if it couldn't be used with the actual AArch64 instructions.
> 
> There was a whole discussion on dot in https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294/13 check out `kparzysz` posts. Essentially Yes we could represent dot this way, but then we would not be able to benefit from the ubquity of the hardware specific dot lowerings that are showing up across gpu and convolution use cases.
> 

Why would using the partial reduction intrinsic stop you from using hardware-specific dot product lowerings for GPUs? The lowering is quite trivial, see [here](https://github.com/llvm/llvm-project/pull/101010). I think it would be best to not introduce another way of doing the same thing.

nikic reviewed (changes requested):
Please create a separate RFC for these intrinsics. I don't think there is a consensus on these intrinsics, and https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294 covers way too much disparate ground (something like "tan" and something like "dot" are entirely separate categories, *especially* if you also want to include integer intrinsics).

farzonl commented:
> Please create a separate RFC for these intrinsics. I don't think there is a consensus on these intrinsics, and https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294 covers way too much disparate ground (something like "tan" and something like "dot" are entirely separate categories, _especially_ if you also want to include integer intrinsics).

@nikic
 The dot case was discussed pretty heavily in the rfc. Would you be open to moving forward if we drop the integer dot intrinsic as a target opcode, but keep the float dot product? We will also put an RFC together for the integer dot intrinsic cases.

@SamTebbs33 For us using an intrisnic better for the float case because you can of specifying rounding modes via vecreduce.  I don't believe `vecreduce.add(mul(zext, zext))` can do that. At least I don't know how you could  but that's a trivial prospect for intrinsics\constraint intrinsics.

Also we are coming at this from different use cases. We want to support language features for now that will be HLSL, but maybe later that includes c++ 20s  `std::inner_product`.   A higher level intrinsic  that is accessible from clang for us is really nice. So this isn't an optimization for us that the vectorizer detects but our basic lowering.

tschuett commented on llvm/include/llvm/Target/GenericOpcodes.td:
```diff
@@ -1088,6 +1088,27 @@ def G_FNEARBYINT : GenericInstruction {
+def G_FDOTPROD : GenericInstruction {
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type0:$src1, type0:$src2);
+  let hasSideEffects = false;
```
dot products take two vectors and return a scalar.
```
f G_FDOTPROD : GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins type1:$src1, type1:$src2);
  let hasSideEffects = false;
```
You have to change the sources to type1 and adapt your legalizer accordingly.

tschuett commented:
Please update the https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MachineVerifier.cpp to protect against misuse.

Commit:
refactor DXIL intrinsic expansion to move erase and replace to the caller

All expansions end with replacing the previous inrinsic with the new
expansion and erasing the old one. By moving this operation to the
caller, these expansion functions can be called in more contexts
and a small amount of duplicated code is consolidated.

Commit:
Fix legalization of dotprod generic opcodes

farzonl commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -284,41 +303,55 @@ static bool expandClampIntrinsic(CallInst *Orig, Intrinsic::ID ClampIntrinsic) {
+  if (Result) {
+    Orig->replaceAllUsesWith(Result);
+    Orig->eraseFromParent();
+  }
```
add a return true to the if block. just keep it simple and leave the `return false;`

tschuett commented on llvm/lib/Target/SPIRV/SPIRVLegalizerInfo.cpp:
```diff
@@ -303,6 +303,14 @@ SPIRVLegalizerInfo::SPIRVLegalizerInfo(const SPIRVSubtarget &ST) {
       allFloatScalarsAndVectors, allIntScalarsAndVectors);
 
+  getActionDefinitionsBuilder(G_FDOTPROD)
+      .legalForCartesianProduct(allFloatScalarsAndVectors,
```
I am not convinced that this is correct. Are you ruling out vectors for the sum type?

tschuett replied:
```
getActionDefinitionBuilder(G_FDOTPROD)
  .legalFor({{s32, s32vector}, {s64, s64vector}, ....});
```

nikic commented:
> > Please create a separate RFC for these intrinsics. I don't think there is a consensus on these intrinsics, and https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294 covers way too much disparate ground (something like "tan" and something like "dot" are entirely separate categories, _especially_ if you also want to include integer intrinsics).
> 
> @nikic The dot case was discussed pretty heavily in the rfc. Would you be open to moving forward if we drop the integer dot intrinsic as a target opcode, but keep the float dot product? We will also put an RFC together for the integer dot intrinsic cases.

There is some discussion in the RFC, but I don't see a consensus on the "dot" intrinsic in particular. I personally haven't found the arguments in favor of it particularly compelling.

This really needs an RFC specific to that intrinsic (class), which includes a clear definition of the semantics of the intrinsics, why it needs to be a intrinsic and how it maps to different hardware (in particular, whether the chosen definition is actually sufficiently portable).

Just looking at the LangRef wordings in this PR, your formulation is not appropriate for target-independent intrinsics. You cannot have a target-independent intrinsic "on a 2-4 element vector of 16-bit or 32-bit floating-point types". The description is also very vague for a floating-point operation. "The arguments are vectors to be elementwise multiplied and then summed" does not tell me *how* the summation occurs. Does it happen in order? As a tree reduction? In unspecified order? Does it use FMA?

Your definitions for sdot and udot also don't really make sense to me -- addition and multiplication are signedness-independent operations, so your sdot and udot are in fact equivalent. Separating these intrinsics only makes sense if they also involve a zero or sign extension, which your formulation does not -- but which would be necessary to support other targets, as pointed out above.

I'll also add that if you are adding target-independent intrinsics, the baseline expectation is that you need to provide full SDAG legalization support for them. I think we've been burned by this enough times that we're not going to accept new target-independent intrinsics that fail to do this.

bogner commented:
> There is some discussion in the RFC, but I don't see a consensus on the "dot" intrinsic in particular. I personally haven't found the arguments in favor of it particularly compelling.
> 
> This really needs an RFC specific to that intrinsic (class), which includes a clear definition of the semantics of the intrinsics, why it needs to be a intrinsic and how it maps to different hardware (in particular, whether the chosen definition is actually sufficiently portable).

@pow2clk @farzonl and I discussed this a bit offline and we've come to agree with you here - while the RFC certainly does point out that there's demand for a generic dot intrinsic I think we were a bit overzealous to try to take it on in this context. What we need in HLSL is a fairly specific subset of the dot operation and it's probably more appropriate for us to just handle that with DirectX and SPIR-V specific intrinsics that do that rather than try to take on defining a fully generic dot operation here.

I'll update the RFC to drop the dot and rsqrt intrinsics, and if we later want to revisit those in a generic way they can be handled on their own merits and pitfalls.

pow2clk commented:
Closing in light of the above. A new PR will capture the DXIL and SPIRV-specific work

pow2clk commented:
Here's the new PR for anyone who wants to keep following along in its altered state: #104656
